collect_app/src/main/java/org/odk/collect/android/entities/DatabaseEntitiesRepository.kt

```diff
@@ -0,0 +1,355 @@
+        if (!listExists(list)) {
+            return emptyList()
+        }
+
+        return databaseConnection.readableDatabase
+            .rawQuery(
+                """
+                SELECT *, i.rowid
+                FROM $list e, ${list}_row_numbers i
+                WHERE e._id = i._id AND e.$property = ?
+                """.trimIndent(),
+                arrayOf(value)
+            ).foldAndClose(emptyList()) { entities, cursor ->
+                entities + mapCursorRowToEntity(
+                    list,
+                    cursor,
+                    cursor.getInt("rowid")
+                )
+            }
+    }
+
+    private fun listExists(list: String): Boolean {
+        return databaseConnection.readableDatabase
+            .query(
+                ListsTable.TABLE_NAME,
+                selection = "${ListsTable.COLUMN_NAME} = ?",
+                selectionArgs = arrayOf(list)
+            ).use { it.count } > 0
+    }
+
+    private fun createList(list: String) {
+        databaseConnection.writeableDatabase.transaction {
+            val contentValues = ContentValues()
+            contentValues.put(ListsTable.COLUMN_NAME, list)
+            insertOrThrow(
+                ListsTable.TABLE_NAME,
+                null,
+                contentValues
+            )
+
+            execSQL(
+                """
+                    CREATE TABLE IF NOT EXISTS $list (
+                        $_ID integer PRIMARY KEY,
+                        ${EntitiesTable.COLUMN_ID} text,
+                        ${EntitiesTable.COLUMN_LABEL} text,
+                        ${EntitiesTable.COLUMN_VERSION} integer,
+                        ${EntitiesTable.COLUMN_TRUNK_VERSION} integer,
+                        ${EntitiesTable.COLUMN_BRANCH_ID} text,
+                        ${EntitiesTable.COLUMN_STATE} integer NOT NULL
+                    );
+                """.trimIndent()
+            )
+
+            execSQL(
+                """
+                CREATE UNIQUE INDEX ${list}_unique_id_index ON $list (${EntitiesTable.COLUMN_ID});
+                """.trimIndent()
+            )
+        }
+    }
+
+    private fun updateProperties(entity: Entity) {
+        entity.properties.map { it.first }.forEach {
+            try {
+                databaseConnection.writeableDatabase.execSQL(
+                    """
+                    ALTER TABLE ${entity.list} ADD $it text;
+                    """.trimIndent()
+                )
+            } catch (e: SQLiteException) {
+                println(e)
+            }
+        }
+    }
+
+    private fun updateRowNumbers() {
+        getLists().forEach {
+            databaseConnection.writeableDatabase.execSQL(
+                """
+                DROP TABLE IF EXISTS ${it}_row_numbers;
+                """.trimIndent()
+            )
+
+            databaseConnection.writeableDatabase.execSQL(
+                """
+                CREATE TABLE ${it}_row_numbers AS SELECT _id FROM $it;
```

Comment:
What is going on here? WH do you need that `_row_numbers` table if it only stores the ids?

Reply:
This is there for the queries in `#getEntities`/`#getById`/`#getAllByProperty`. I think the best way to understand what's going on is to think about the `InMem...` implementation of `#getAllByProperty` which basically does:

```kotlin
entities.filter { it.list == list }.mapIndexed { index, entity ->
    Entity.Saved(
        entity.list,
        entity.id,
        entity.label,
        entity.version,
        entity.properties,
        entity.state,
        index,
        entity.trunkVersion,
        entity.branchId
    )
}.filter { entity ->
    entity.properties.any { (first, second) -> first == property && second == value
}
```

Here we're using the position in the list as the `index` for each entity and then filtering. In SQL that's a little more tricky as there's no simple way to get the "position" of a row in a table. Newer versions of SQLite actually have a `row_number()` function to handle this, but we can't use that on the versions of Android we support. One solution is to use a nested query like I do in [this commit](https://github.com/getodk/collect/pull/6290/commits/ea61916daf1e33f6b4b9bda7e45fc41c1a69a268), but that was very slow - SQLite probably ends up with something close to an "n squared" query to execute that. I found that creating an "index" table that maps IDs to row numbers was really fast to query and to recreate on any `#save` so I've ended up with this.

The core need here (that the tests enforce) is that `index`/`id` is consistent between queries (until the data changes) and that there are no gaps in `index` so that it works for multiplicity (how XPath will address the entities). Definitely let me know if you can think of other solutions though! It's a frustratingly tricky one.

Reply:
Ok now I understand but it definitely took me too much time to wrap my head around this concept. Please document the purpose of those `_row_numbers` tables in detail. It's also worth adding that once our min SDK version is bumped (API 29?) we will be able to use the `row_number()` function.

Reply:
>  It's also worth adding that once our min SDK version is bumped (API 29?) we will be able to use the row_number() function.

I wanted to initially focus on experimenting with solutions that work on every device, but I'm also now curious about the performance we'd actually get with `row_number()`. I'll add an issue to investigate later.